File: MapboxVelocity/MapboxVelocity/Features/Map/RideAnimator.swift
//
//  RideAnimator.swift
//  MapboxVelocity
//
//  Created by Shaikat on 4/26/26.
//

// flow
/*
 1. User picks pickup → sets a pin on map
 2. User picks destination → calls Directions API, draws the route polyline
 3. User selects vehicle + taps Book → tripViewModel.startRide() sets state to .riding
 4. MapViewController sees .riding state → calls rideAnimator.start(along: routeCoordinates)
 5. Every 100ms, RideAnimator fires the next coordinate → MapViewController moves a car/bike annotation along the polyline
 6. When the last coordinate is reached → onCompletion fires → tripViewModel.endRide() → state becomes .rideEnded
 */

import Foundation
import CoreLocation

// This is to mimic the ride sharing flow

final class RideAnimator {
    // Called every tick with the current coordinate and bearing toward next point
    var onCoordinateUpdate: ((_ coordinate: CLLocationCoordinate2D, _ bearing: Double) -> Void)?

    // Called when the animation finishes
    var onCompletion: (() -> Void)?

    private var timer: Timer?
    private var coordinates: [CLLocationCoordinate2D] = []
    private var currentIndex: Int = 0
    // Seconds between each step — controls visual speed
    let stepInterval: TimeInterval = 0.10

    func start(along coordinates: [CLLocationCoordinate2D]) {
        guard coordinates.count > 1 else { return }
        stop()
        self.coordinates = coordinates
        self.currentIndex = 0

        timer = Timer.scheduledTimer(withTimeInterval: stepInterval,
                                     repeats: true) { [weak self] _ in
            self?.tick()
        }
    }

    func stop() {
        timer?.invalidate()
        timer = nil
    }

    private func tick() {
        guard currentIndex < coordinates.count else {
            stop()
            onCompletion?()
            return
        }
        let current = coordinates[currentIndex]
        // Bearing toward next point, or same as last segment if at the end
        let next = currentIndex + 1 < coordinates.count ? coordinates[currentIndex + 1] : current
        let bearing = current.bearing(to: next)
        onCoordinateUpdate?(current, bearing)
        currentIndex += 1
    }
}

File: MapboxVelocityAndroid/app/src/main/java/com/mapboxvelocity/Features/Map/RideAnimator.kt
package com.mapboxvelocity.Features.Map

import android.os.Handler
import android.os.Looper
import com.mapbox.geojson.Point
import com.mapboxvelocity.Core.Extensions.bearing

// Mirrors iOS RideAnimator:
// - Fires every 100ms (stepInterval)
// - Calls onCoordinateUpdate with current position + bearing toward next point
// - Calls onCompletion when last coordinate is reached
class RideAnimator {

    var onCoordinateUpdate: ((point: Point, bearing: Double) -> Unit)? = null
    var onCompletion: (() -> Unit)? = null

    val stepInterval: Long = 100L // ms — matches iOS 0.10s

    private val handler = Handler(Looper.getMainLooper())
    private var coordinates: List<Point> = emptyList()
    private var currentIndex = 0

    private val ticker = object : Runnable {
        override fun run() {
            tick()
        }
    }

    fun start(coordinates: List<Point>) {
        if (coordinates.size < 2) return
        stop()
        this.coordinates = coordinates
        this.currentIndex = 0
        handler.post(ticker)
    }

    fun stop() {
        handler.removeCallbacks(ticker)
    }

    private fun tick() {
        if (currentIndex >= coordinates.size) {
            stop()
            onCompletion?.invoke()
            return
        }
        val current = coordinates[currentIndex]
        val next = if (currentIndex + 1 < coordinates.size) coordinates[currentIndex + 1] else current
        val bearing = current.bearing(next)
        onCoordinateUpdate?.invoke(current, bearing)
        currentIndex++
        handler.postDelayed(ticker, stepInterval)
    }
}
